### PR TITLE
Cleanup RMC Flags

### PR DESCRIPTION
**Pull request by vecchiot-aws**
### Description of changes: 

This PR cleans up the flags in `rmc` and `cargo-rmc` by unifying them into a shared file, adding group organization, and descriptions for each flag.

### Testing:

* How is this change tested? Existing regression suite.

* Is this a refactor change? Yes.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: scripts/rmc_flags.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+def add_loudness_flags(make_group, add_flag):
+    group = make_group("Loudness flags", "Determine how much textual output to produce.")
+    add_flag(group, "--debug", action="store_true", 
+        help="Produce full debug information")
+    add_flag(group, "--verbose", "-v", action="store_true",
+        help="Output processing stages and commands, along with minor debug information")
+    add_flag(group, "--quiet", "-q", action="store_true",
+        help="Produces no output, just an exit code and requested artifacts. Overrides --verbose")
+
+def add_um_flags(make_group, add_flag):
```
Comment (nchong-at-aws) [inline scripts/rmc_flags.py:16]: Maybe proof_harness_flags?

File: scripts/rmc_flags.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+def add_check_flags(make_group, add_flag):
+    group = make_group("Check flags", "Disable some or all default checks.")
+    add_flag(group, "--no-default-checks", action="store_true", help="Disable all default checks")
+    add_flag(group, "--no-memory-safety-checks", action="store_true", help="Disable default memory safety checks")
+    add_flag(group, "--no-overflow-checks", action="store_true", help="Disable default overflow checks")
+    add_flag(group, "--no-unwinding-checks", action="store_true", help="Disable default unwinding checks")
+
+def add_visualizer_flags(make_group, add_flag):
+    group = make_group("Visualizer flags", "Generate an HTML-based UI for the generated RMC report.")
```
Comment (nchong-at-aws) [inline scripts/rmc_flags.py:42]: Consider pointing to the cbmc-viewer docs for more info

File: scripts/rmc_flags.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    add_flag(group, "--srcdir", default=".") # TODO: help?
+    add_flag(group, "--wkdir", default=".") # TODO: help?
+    add_flag(group, "--visualize", action="store_true",
+        help="Generate visualizer report; open report/html/index.html")
+
+def add_other_flags(make_group, add_flag):
+    group = make_group("Other flags")
+    add_flag(group, "--allow-cbmc-verification-failure", action="store_true",
+        help="Do not produce error retcode on CBMC verification failure")
```
Comment (nchong-at-aws) [inline scripts/rmc_flags.py:51]: retcode -> return code

File: scripts/rmc_flags.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    add_flag(group, "--mangler", default="v0",
+        help="Change what mangler is used by the Rust compiler")
+    add_flag(group, "--dry-run", action="store_true", 
+        help="Print commands instead of running them")
+
+def add_developer_flags(make_group, add_flag):
+    group = make_group("Developer flags", "These are generally meant for use by RMC developers, and are not necessarily stable.")
```
Comment (nchong-at-aws) [inline scripts/rmc_flags.py:58]: Remove necessarily

**Review by nchong-at-aws — CHANGES_REQUESTED** [review]
Good clean-up. Some suggested/requested changes on naming.

File: scripts/rmc_flags.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    add_flag(group, "--cbmc-args", nargs=argparse.REMAINDER, default=[], 
+                        help="Pass through directly to CBMC; must be the last flag")
+
+def add_flags(
+    parser, 
+    exclude_flags=[],
+    exclude_groups=[]):
```
Comment (bdalrhm) [inline scripts/rmc_flags.py:65]:
```suggestion
def add_flags(parser, exclude_flags=[], exclude_groups=[]):
```
Consider using a formatter (like `autopep`)

File: scripts/cargo-rmc [hunk 6]
```diff
@@ -20,28 +21,12 @@ def main():
         del sys.argv[1]
 
     parser = argparse.ArgumentParser(description="Verify a Rust crate")
```
Comment (bdalrhm) [inline scripts/cargo-rmc:23]:
```suggestion
    parser = argparse.ArgumentParser(prog="cargo rmc", description="Verify a Rust crate")
```
`cargo-rmc` script is usually called as `cargo rmc`.

**Review by bdalrhm — COMMENTED** [review]
Appreciate the help messages 🙂

File: scripts/cargo-rmc [hunk 7]
```diff
@@ -56,6 +41,12 @@ def main():
     if not rmc.dependencies_in_path():
         return 1
 
+    if args.gen_symbols:
+        print("WARNING: --gen-symbols is not supported for cargo-rmc.")
+
+    if args.allow_cbmc_verification_failure:
+        print("WARNING: --allow-cbmc-verification-failure is not supported for cargo-rmc.")
```
Comment (avanhatt) [inline scripts/cargo-rmc:48]: nit: can we fail here instead of warn? Warnings are easy to overlook

File: scripts/cargo-rmc [hunk 8]
```diff
@@ -56,6 +41,12 @@ def main():
     if not rmc.dependencies_in_path():
```
Comment (avanhatt) [inline scripts/cargo-rmc:41]: I know this isn't your change, but should this give a more helpful message?

File: scripts/rmc [hunk 9]
```diff
@@ -16,29 +17,12 @@ CBMC_VERIFICATION_FAILURE_EXIT_CODE = 10
...
-    parser.add_argument("--wkdir", default=".")
-    parser.add_argument("--no-default-checks", action="store_true", help="Disable all default checks")
-    parser.add_argument("--no-memory-safety-checks", action="store_true", help="Disable default memory safety checks")
-    parser.add_argument("--no-overflow-checks", action="store_true", help="Disable default overflow checks")
-    parser.add_argument("--no-unwinding-checks", action="store_true", help="Disable default unwinding checks")
-    parser.add_argument("--dry-run", action="store_true", help="Print commands instead of running them")
-    parser.add_argument("--cbmc-args", nargs=argparse.REMAINDER,
-                        default=[], help="Pass through directly to CBMC")
+
+    input_group = parser.add_argument_group("Input", "You can pass in the rust file positionally or with the --input flag.")
```
Comment (avanhatt) [inline scripts/rmc:21]: Should this specify in the last position?

Comment (vecchiot-aws) [reply scripts/rmc:21]: It doesn't have to be in the last position; actually putting it first can avoid issues with flags that take multiple arguments. The only thing that *really* matters is `--cbmc-args` last, or else some RMC flags might be interpretted as CBMC flags.

File: scripts/cargo-rmc (hunk 8 above)
Comment (vecchiot-aws) [reply scripts/cargo-rmc:41]: The error is printed by `dependencies_in_path`, so we don't need to print anything here.

File: scripts/rmc_flags.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        add_group(make_group, add_flag)
+
+    # Error if any excluded flags/groups don't exist.
+    extra_groups = set(exclude_groups) - excluded_groups
+    if extra_groups:
+        print(f"ERROR: Attempt to exclude groups which don't exist: {extra_groups}")
+
+    extra_flags = set(exclude_flags) - excluded_flags
+    if extra_flags:
+        print(f"ERROR: Attempt to exclude flags which don't exist: {extra_flags}")
```
Comment (vecchiot-aws) [inline scripts/rmc_flags.py:131]: Raise an error instead; "Error setting up command line parser..." + link to issue page for users to report bug.

File: scripts/rmc_flags.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    group = make_group(
+        "Artifact flags", "Produce artifacts in addition to a basic RMC report.")
+    add_flag(group, "--target-dir", default=default_target, metavar="DIR",
+             help=f"Directory for all generated artifacts; defaults to \"{default_target}\"")
+    add_flag(group, "--keep-temps", action="store_true",
+             help="Keep temporary files generated throughout RMC process")
+    add_flag(group, "--gen-c", action="store_true",
+             help="Generate C file equivalent to inputted program")
+    add_flag(group, "--gen-symbols", action="store_true",
+             help="Generate a goto symbol table")
```
Comment (danielsn) [inline scripts/rmc_flags.py:43]: sort

**Review by danielsn — APPROVED** [review]

**Review by nchong-at-aws — APPROVED** [review]

**Pull request merged**